File: HW2/src/shift_encryption.cpp
#include <cstdio>
#include "shift_encryption.h"
#include "error.h"
#include "text.h"

ShiftEncryption::~ShiftEncryption() {
    delete[] encrypted_text_;
}

void ShiftEncryption::Input(FILE *file) {
    if (fscanf(file, "%u ", &shift_) == 0) {
        Error("Incorrect shift");
    }
}

void ShiftEncryption::RandomEncryption(size_t min_param, size_t max_param) {
    shift_ = rand() % (max_param - min_param + 1) + min_param;
}

void ShiftEncryption::Output(FILE *file) const {
    Text::Output(file);
    fprintf(file, "Shift: %d; Result: \"%s\"\n", shift_, encrypted_text_);
}

void ShiftEncryption::Encrypt() {
    encrypted_text_ = new unsigned char[length_ + 1];
    int shift = shift_ % 8;
    for (size_t index = 0; index < length_; ++index) {
        unsigned char symbol = string_[index];
        symbol = (symbol << shift) | (symbol >> (8 - shift));
        encrypted_text_[index] = (char)symbol;
    }
    encrypted_text_[length_] = '\0';
}

File: HW3/src/shift_encryption.py
import random
from functools import cached_property
from typing import TextIO

from src.text import Text
from src.utils import read_until_space


class ShiftEncryption(Text):
    CODE_POINT_SIZE = 7

    def __init__(self, text: str, shift: int):
        super().__init__(text)
        self.shift = shift

    @classmethod
    def from_file(cls, text: str, input_file: TextIO) -> Text:
        try:
            shift = int(read_until_space(input_file))
        except Exception:
            raise ValueError("Incorrect shift")
        return cls(text, shift)

    @classmethod
    def random_encryption(cls, text: str) -> Text:
        return cls(text, random.randint(0, 20))

    def output(self, output_file: TextIO) -> None:
        super(ShiftEncryption, self).output(output_file)
        print(f'Shift: {self.shift}; Result: "{self.encrypted_text}"',
              file=output_file)

    @cached_property
    def encrypted_text(self) -> str:
        left_shift = self.shift % self.CODE_POINT_SIZE
        right_shift = self.CODE_POINT_SIZE - self.shift % self.CODE_POINT_SIZE
        return ''.join(chr(((ord(symbol) << left_shift) | (
                    ord(symbol) >> right_shift)) & (
                                       (1 << self.CODE_POINT_SIZE) - 1)) for
                       symbol in self.text)
